Simperium/src/main/java/name/fraser/neil/plaintext/diff_match_patch.java

```diff
@@ -1524,6 +1611,27 @@ public String diff_toDelta(LinkedList<Diff> diffs) {
               "Negative number in diff_fromDelta: " + param);
         }
         String text;
+        // some objective-c versions of the library produced patches with
+        // (null) in the place where surrogates were split across diff
+        // boundaries. if we leave those in we'll be stuck with a
+        // high-surrogate (null) low-surrogate pattern that will break
+        // deeper in the library or consuming application. we'll "fix"
+        // these by dropping the (null) and re-joining the surrogate halves
+        if (x + 2 < tokens.length &&
+            Character.isHighSurrogate(text1.charAt(pointer + n - 1)) &&
+            tokens[x + 1].substring(1).equals("(null)") &&
+            Character.isLowSurrogate(text1.charAt(pointer + n))) {
+          n -= 1;
+          tokens[x + 1] = "+";
+          int m;
+          try {
+            m = Integer.parseInt(tokens[x + 2].substring(1));
+          } catch (NumberFormatException e) {
+            throw new IllegalArgumentException(
+                    "Invalid number in diff_fromDelta: " + tokens[x + 2].substring(1), e);
+          }
+          tokens[x + 2] = tokens[x + 2].charAt(0) + String.valueOf(m + 1);
```

Comment:
it's my understanding that @jleandroperez did not incorporate this into the iOS library, meaning that while I'm pretty confident in it, it's untested _in practice_

Reply:
While testing these changes, the `(null)` part was not removed from a corrupted note with the `high-surrogate (null) low-surrogate` pattern.  According to the comment above this block, this code should be attempting to repair notes with that corrupted content, right?

Here is how I tested it.
1. Create corrupted note using another platform.
2. Launch app and open corrupted note.
3. Edit corrupted note before or after corruption.

Reply:
The `ChannelTest.testRequestObjectForInvalidChange()` test is failing when running `./gradlew connectedCheck` on this branch.  Interestingly, removing the block in question here makes the test pass.

Reply:
@theck13 @dmsnell reason why I didn't add the (repair) part of the fix, is that:

- Once a string is badly split, `(null)` is generated by the iOS SDK.
- And we do get data loss (the `(null)` is sent over the wire).

Meaning that... once the issue shows up, it's irrecoverable. That is, as long as the badly split string is generated ios-side.

Now, that being said, I couldn't trigger any test case in which a malformed string could be patched up, other than unit tests. For that reason, for the time being, I've opted to change as little as possible in DMP (it's been quite stable for the past couple releases!)

Reply:
That's fine @jleandroperez - the patch doesn't attempt to recover lost data - it attempts to undo the change vs. leave the unexpected `(null)` in there - something I'm willing to acknowledge is more of a "how do you prefer to proceed when things are broken" question than "this is the right choice"

that being said I'm surprised that tests failed and I'll be looking into the code to try and ifgure out what's up, also surprised that it didn't remove the `(null)`

my guess is that I got something wrong in the Java code but I do have at least one test confirming the behavior in the `diff-match-patch` library PR

Reply:
> Create corrupted note using another platform.
> Launch app and open corrupted note.

After investigating why we had unexpected corruption on Android and pinpointing `sqlite` as the perpetrator, I believe that it makes sense that the reason we're not seeing the `(null)` issue fixed here is because we're not seeing the corrupted patch come in. This fix will only adjust patching which when applied would produce that bug and it appears like we're starting with a copy of the note from the server which won't present the buggy patch.

In other words, in those steps I don't think we're reproducing the `(null)` bug.